### ISSUE TITLE
become a subreaper

### ISSUE THREAD

**Issue by lilydjwg** [i3/i3#4274]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior
`exec`ed processes become children of pid 1.

## Desired Behavior

`exec`ed processes become children of i3, even they fork and fork.

A call like `prctl(PR_SET_CHILD_SUBREAPER, 1)` will make a process become a subreaper, and all descendant processes will become children of it when their parent process dies. This will make the process tree clear to understand and easier to track for the user. systemd has been using this.

I'm also using this with zsh, Awesome and dcron.

For zsh, it's rather simple: a module that does the call just works.

For Awesome, I need to take care to actually reap (`waitpid`) child processes, and `signal(SIGCHLD, SIG_IGN)` to ignore future child zombie processes. I've done this via a lua module written in C and loaded it in its configuration file.

For dcron, I have to modify the source code and change some `waitpid` logic to make it work correctly.

I can try to make a pull request if such a feature is going to be merged.

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
</pre>

<pre>
- Linux Distribution & Version: Arch Linux
- Are you using a compositor (e.g., xcompmgr or compton):
</pre>

**Comment by psychon** [comment]
> For Awesome, I need to take care to actually reap (waitpid) child processes, and signal(SIGCHLD, SIG_IGN) to ignore future child zombie processes. I've done this via a lua module written in C and loaded it in its configuration file.

Congratulations, you just broke awesomeWM (it can no longer wait for child processes and provide their exit status). Since the following commit, AwesomeWM will reap children itself (it calls `waitpid(-1)`):
```
commit 6c559e188c2aa3b6e3777c2bd81d0219967eeb38
Author: Uli Schlachter <psychon@znc.in>
Date:   Tue Aug 21 15:30:07 2018 +0200

    Handle SIGCHLD ourselves instead of through GLib
```
(The above commit has a longer commit message; it was explicitly done this way to avoid zombies accumulating.)

**Comment by Trumeet** [comment]
I had the same problem, and I just made a simple wrapper to start i3 with `PR_SET_CHILD_SUBREAPER` flag:

```c
int main(int argc, char **argv) {
        assert(argc);
        if (argc == 1)
                errx(64, "Usage: %s /path/to/binary [arg1] [arg2] ...",
                                argv[0]);
        if (prctl(PR_SET_CHILD_SUBREAPER, 1) < 0)
                err(errno, "prctl(PR_SET_CHILD_SUBREAPER)");
        if (execve(argv[1], &argv[1], environ))
                err(errno, "exec(%s)", argv[1]);
}
```

Compile it and execute as `/path/to/output /usr/bin/i3` will do the trick. It will exec(2) i3, so no additional processes will be forked.

**Comment by markstos** [comment]
If this were implemented, would restarting i3 cause all apps to be killed?

**Comment by markstos** [comment]
Is this a dupe of [#4073 Make all processes a sub-process of i3](https://github.com/i3/i3/issues/4073) ?

**Comment by lilydjwg** [comment]
> If this were implemented, would restarting i3 cause all apps to be killed?

No. They will be reparented to another process if current parent dies.

> Is this a dupe of #4073 ?

Somewhat yes. #4073 is the desired result; this is a suggested implementation.
